java/core/src/main/java/com/google/protobuf/CodedOutputStream.java

```diff
@@ -1348,31 +1348,36 @@ public final void writeUInt32NoTag(int value) throws IOException {
       if (HAS_UNSAFE_ARRAY_OPERATIONS
           && !Android.isOnAndroidDevice()
           && spaceLeft() >= MAX_VARINT32_SIZE) {
-        if ((value & ~0x7F) == 0) {
-          UnsafeUtil.putByte(buffer, position++, (byte) value);
-          return;
-        }
-        UnsafeUtil.putByte(buffer, position++, (byte) (value | 0x80));
-        value >>>= 7;
-        if ((value & ~0x7F) == 0) {
-          UnsafeUtil.putByte(buffer, position++, (byte) value);
-          return;
-        }
-        UnsafeUtil.putByte(buffer, position++, (byte) (value | 0x80));
-        value >>>= 7;
-        if ((value & ~0x7F) == 0) {
-          UnsafeUtil.putByte(buffer, position++, (byte) value);
-          return;
-        }
-        UnsafeUtil.putByte(buffer, position++, (byte) (value | 0x80));
-        value >>>= 7;
-        if ((value & ~0x7F) == 0) {
+        int position = this.position;
```

Comment:
Interesting. Two things of note:

1. I assume CodedOutputStream is not expected to be thread-safe, and it doesn't look like it is now, but I need to verify that and think about whether position might be updated asynchronously.

2. This whole block doesn't run on android so the relevant benchmarks are for server.

'Thanks!

Reply:
The existing behavior is definitely not thread-safe, and it wouldn't make much sense to use simultaneous writes on a single output stream. Any extra information written while serializing a message would break at least the length delimiters and add fields that the parser may not be expecting.